diesel_codegen/src/associations.rs

```diff
@@ -0,0 +1,103 @@
+use syn;
+use quote;
+
+use model::{Model, infer_association_name};
+use util::str_value_of_meta_item;
+
+pub fn derive_associations(input: syn::MacroInput) -> quote::Tokens {
+    let mut derived_associations = Vec::new();
+    let model = t!(Model::from_item(&input, "Associations"));
+
+    for attr in &input.attrs {
+        if attr.name() == "has_many" {
+            let options = t!(build_association_options(attr, "has_many"));
+            derived_associations.push(expand_has_many(&model, options))
+        }
+
+        if attr.name() == "belongs_to" {
+            let options = t!(build_association_options(attr, "belongs_to"));
+            derived_associations.push(expand_belongs_to(&model, options))
+        }
+    }
+
+    quote!(#(derived_associations)*)
+}
+
+fn expand_belongs_to(model: &Model, options: AssociationOptions) -> quote::Tokens {
+    let parent_struct = options.name;
+    let struct_name = &model.name;
+
+    let foreign_key_name = options.foreign_key_name.unwrap_or_else(||
+        to_foreign_key(&parent_struct.as_ref()));
+    let child_table_name = model.table_name();
+    let fields = model.attrs.clone();
+
+    quote!(BelongsTo! {
+        (
+            struct_name = #struct_name,
+            parent_struct = #parent_struct,
+            foreign_key_name = #foreign_key_name,
+            child_table_name = #child_table_name,
+        ),
+        fields = [#(fields)*],
+    })
+}
+
+fn expand_has_many(model: &Model, options: AssociationOptions) -> quote::Tokens {
+    let parent_table_name = model.table_name();
+    let child_table_name = options.name;
+    let foreign_key_name = options.foreign_key_name.unwrap_or_else(||
+        to_foreign_key(&model.name.as_ref()));
+    let fields = model.attrs.clone();
```

Comment:
/cc @dtolnay It looks like this is an issue in quote. I can't actually use borrowed data here. Since you're always putting at addr-of in front of the value before iterating over it, it's effectively `T where for<'a> &'a T: IntoIterator`, which is only true for `Vec`. It looks like *with* a separator you are taking ownership of the value though. Worth always taking ownership of the iterator? Do you have a better suggestion?